File: Python/47. Permutations II.py
class Solution:
    def permuteUnique(self, nums: list[int]) -> list[list[int]]:
        result = []
        def backtrack(n, nums):
            if n == len(nums):
                result.append(nums.copy())
                return 
            
            unique = set()
            for i in range(n, len(nums)):
                if nums[i] not in unique:
                    unique.add(nums[i])
                    nums[i], nums[n] = nums[n], nums[i]

                    backtrack(n+1, nums)

                    nums[n], nums[i] = nums[i], nums[n]                    
            return 

        backtrack(0, nums)
        return result

File: Javascript/47. Permutations II.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function(nums) {
    return backtrack(0, nums, [])
};

var backtrack = function(s, nums, ans) {
    if (s == nums.length) {
        ans.push([...nums])
        return ans
    }
    
    let sets = new Set()
    for (let i = s; i < nums.length; i++) {
        if (!sets.has(nums[i])) {
            sets.add(nums[i])
            temp = nums[i]
            nums[i] = nums[s]
            nums[s] = temp

            ans = backtrack(s+1, nums, ans)
            
            temp = nums[i]
            nums[i] = nums[s]
            nums[s] = temp
        }
    }
    return ans
}
